core/src/main/java/org/pentaho/platform/engine/services/connection/datasource/dbcp/NonPooledOrJndiDatasourceService.java

```diff
@@ -50,10 +48,8 @@ protected DataSource retrieve( String dsName ) throws DBDatasourceServiceExcepti
       if ( databaseConnection != null && !databaseConnection.getAccessType().equals( DatabaseAccessType.JNDI ) ) {
         ds = resolveDatabaseConnection( databaseConnection );
         // Database does not have the datasource, look in jndi now
-      } else if ( databaseConnection == null ) {
-        ds = getJndiDataSource( dsName );
       } else {
-        ds = getJndiDataSource( databaseConnection.getDatabaseName() );
+        ds = getJndiDataSource( dsName );
```

Comment:
Not sure I agree with removing the getJndiDataSource( databaseConnection.getDatabaseName() ) part. I think this will cause regressions. Suggest trying the getJndiDataSource( dsName ) and if that fails to return a datasource, then use the other.